datasource/ethereum/src/ethereum_adapter.rs

```diff
@@ -1084,4 +1009,156 @@ where
+            BlockFinality::Final(block) => Box::new(
+                self.blocks_with_triggers(
+                    logger,
+                    chain_store,
+                    subgraph_metrics,
+                    block.number(),
+                    block.number(),
+                    log_filter.clone(),
+                    call_filter.clone(),
+                    block_filter.clone(),
+                )
+                .map(|blocks| {
+                    assert!(blocks.len() <= 1);
+                    blocks
+                        .into_iter()
+                        .next()
+                        .unwrap_or(EthereumBlockWithTriggers::new(vec![], ethereum_block))
+                }),
+            )
+                as Box<dyn Future<Item = _, Error = _> + Send>,
+            BlockFinality::NonFinal(full_block) => Box::new(future::ok({
+                let mut triggers = Vec::new();
+                triggers.append(&mut parse_log_triggers(
+                    log_filter,
+                    &full_block.ethereum_block,
+                ));
+                triggers.append(&mut parse_call_triggers(call_filter, &full_block));
+                triggers.append(&mut parse_block_triggers(block_filter, &full_block));
+                EthereumBlockWithTriggers::new(triggers, ethereum_block)
+            })),
+        })
+    }
+
+    /// Load Ethereum blocks in bulk, returning results as they come back as a Stream.
+    fn load_blocks(
+        &self,
+        logger: Logger,
+        chain_store: Arc<dyn ChainStore>,
+        block_hashes: HashSet<H256>,
+    ) -> Box<dyn Stream<Item = LightEthereumBlock, Error = Error> + Send> {
+        // Search for the block in the store first then use json-rpc as a backup.
+        let mut blocks = chain_store
+            .blocks(block_hashes.iter().cloned().collect())
+            .map_err(|e| error!(&logger, "Error accessing block cache {}", e))
+            .unwrap_or_default();
+
+        let missing_blocks = Vec::from_iter(
+            block_hashes
+                .into_iter()
+                .filter(|hash| !blocks.iter().any(|b| b.hash == Some(*hash))),
+        );
+
+        // Return a stream that lazily loads batches of blocks.
+        debug!(logger, "Requesting {} block(s)", missing_blocks.len());
+        Box::new(
+            self.load_blocks_rpc(logger.clone(), missing_blocks.into_iter().collect())
+                .collect()
+                .map(move |new_blocks| {
+                    if let Err(e) = chain_store.upsert_thin_blocks(new_blocks.clone()) {
+                        error!(logger, "Error writing to block cache {}", e);
+                    }
+                    blocks.extend(new_blocks);
+                    blocks.sort_by_key(|block| block.number);
+                    stream::iter_ok(blocks)
+                })
+                .flatten_stream(),
+        )
+    }
+
+    /// Reorg safety: `to` must be a final block.
+    fn block_range_to_ptrs(
```

Comment:
Is this wrapper around `load_block_ptrs_rpc` necessary at all?

Reply:
It would be in the same "layer" as `load_blocks`. If we ever figure out how to cache this data to the DB, this is where we'd handle that.